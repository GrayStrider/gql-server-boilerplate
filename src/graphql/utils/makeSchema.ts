import {IExecutableSchemaDefinition, makeExecutableSchema} from 'graphql-tools'
import {importSchema} from 'graphql-import'

export default async function makeSchema<TContext> (schemaPath: string, params: Omit<IExecutableSchemaDefinition, 'typeDefs'>) {
	
	const typeDefs = await importSchema(schemaPath, {})
	return makeExecutableSchema<TContext>({...params, typeDefs})
	
}

