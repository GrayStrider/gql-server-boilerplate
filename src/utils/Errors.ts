import {truncate} from 'lodash'
import {AnyObject} from 'tsdef'
import {ApolloError} from 'apollo-server-errors'


// eslint-disable-next-line @typescript-eslint/interface-name-prefix
interface IExpectedError {
	details?: AnyObject | string
}

function makeCustomError (
	code: ErrorCodes, defaultMessage: string, defaultDetails?: AnyObject | string
) {
	
	return class ExpectedError extends ApolloError implements IExpectedError {
		
		public details?: AnyObject | string
		
		constructor (message?: string, details?: AnyObject | string) {
			
			super(message ?? defaultMessage, code)
			this.details = details ?? defaultDetails
			
		}
		
	}
	
}

enum ErrorCodes {
	'VALIDATION_ERROR' = 'VALIDATION_ERROR',
	'NOT_FOUND' = 'NOT_FOUND',
	'UNATHORIZED' = 'UNATHORIZED'
}

const Errors = {
	Validation: makeCustomError(ErrorCodes.VALIDATION_ERROR, 'Unspecified validation error'),
	NotFound: makeCustomError(ErrorCodes.NOT_FOUND, 'Object not found'),
	Unathorized: makeCustomError(ErrorCodes.UNATHORIZED, 'Unathorized to perform requested action'),
	InvalidCredentials: makeCustomError(ErrorCodes.UNATHORIZED, 'Invalid credentials provided'),
}

const userNotFoundError = (id: string) =>
	new Errors.NotFound(`User ${truncate(id, {length: 10})} not found`)

export default Errors
export {ErrorCodes, userNotFoundError, makeCustomError, IExpectedError}
