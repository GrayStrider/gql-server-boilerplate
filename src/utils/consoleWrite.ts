export default function consoleWrite (message?: unknown, ...optionalParams: unknown[]) {

	process.stdout.write(`${JSON.stringify(message)}`)
	
	optionalParams.forEach(msg => process.stdout.write(`${JSON.stringify(msg)}`))
	
}
