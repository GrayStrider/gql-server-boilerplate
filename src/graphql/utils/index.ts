import makeSchema from '@/graphql/utils/makeSchema'

export {
	makeSchema,
}
