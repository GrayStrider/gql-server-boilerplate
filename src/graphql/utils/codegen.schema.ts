import {promises} from 'fs'
import {filter, test as regexpTest, isEmpty} from 'ramda'
import {buildSchema, GraphQLSchema} from 'graphql'
import {mergeSchemas} from 'graphql-tools'
import {bb} from '@/utils'

const {readdir, readFile} = promises

const PATH = 'src/graphql/generated/'

async function main (): Promise<GraphQLSchema> {
	
	const files = await readdir(PATH)
		.then(filter(regexpTest(/.graphql$/u)))
	
	if (isEmpty(files)) {

		const defaultSchema = `
        type Query {
            placeholder: String
        }
        
        type Mutation {
        		placeholder: String
        }
		`
		return buildSchema(defaultSchema)
	
	}
	
	const schemas = await bb.map(files,
		async fileName =>
			readFile(PATH + fileName, 'utf8'))
	
	return mergeSchemas({schemas})
	
}


export default main()
	.then(res => res)
