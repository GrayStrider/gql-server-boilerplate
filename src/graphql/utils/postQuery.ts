import {ASTNode, print} from 'graphql'
import {request} from 'graphql-request'
import {Variables, GraphQLResponse} from 'graphql-request/dist/src/types'
import {AnyObject} from 'tsdef'
import {GQL_URL} from '@config'
import {Mutation, Query} from '@/graphql/generated/typings'
import {flattenGQLResponse} from '@/utils'


async function gqlRequest<T> (query: ASTNode, variables?: Variables, url?: string): Promise<T>
async function gqlRequest<T, K> (query: ASTNode, variables?: Variables, url?: string): Promise<[T, K]>
async function gqlRequest<T, K, U> (query: ASTNode, variables?: Variables, url?: string): Promise<[T, K, U]>


async function gqlRequest<T, K, U> (
	query: ASTNode, variables?: Variables, url: string = GQL_URL
) {
	
	const res: GraphQLResponse = await request(
		url, print(query), variables
	)
	
	return flattenGQLResponse(res)
	
}


type KeyofQuery = keyof Omit<Query, '__typename'>
type KeyofMutation = keyof Omit<Mutation, '__typename'>

async function gqlreq (
	type: 'query' | 'mutation', obj: KeyofMutation | KeyofQuery, query: ASTNode, variables?: AnyObject, url: string = GQL_URL
) {
	
	const res = await request(
		url, print(query), variables
	)
	return flattenGQLResponse(res)
	
}

export {gqlRequest, gqlreq}
