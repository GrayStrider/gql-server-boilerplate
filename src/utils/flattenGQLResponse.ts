import {keys, head, values} from 'ramda'
import {GraphQLResponse} from 'graphql-request/dist/src/types'

/**
 * Refactored to be only used in GraphQL responses
 *
 * Yes the function actually returns T | V (input),
 * but the assumption here is that the consumer already knows what kind of data
 * he's gonna get, so that's mosly for convinience, and
 * you totally can specify incorrect types and get a runtime "undefined".
 * It's meant for use in tests, anyway.
 * Too much trouble handling these edge cases
 * @param response
 */
export default function flattenGQLResponse<T extends unknown> (response: GraphQLResponse): T {
	
	if (response.errors && response.errors.length > 0)
		throw new Error(response.errors[0].message)
	
	
	const {data} = response
	
	if (!data) return response as T
	if (keys(data).length > 1) return data as T
	
	/**
	 * Return the first (head)
	 * value (values) of the response
	 * same as Object.values(response)[0]
	 */
	return head(values(data)) as T
	
}

