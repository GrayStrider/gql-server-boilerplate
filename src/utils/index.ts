import signale from 'signale'
import * as RD from 'ramda-adjunct'
import {Promise as bb} from 'bluebird'
import gql from 'graphql-tag'
import Errors from '@/utils/Errors'
import log from '@/utils/log'
import flattenGQLResponse from '@/utils/flattenGQLResponse'
import warn from '@/utils/warn'
import sleep from '@/utils/sleep'
import consoleWrite from '@/utils/consoleWrite'

signale.config({
	displayScope: true,
	displayBadge: false,
	displayDate: false,
	displayFilename: true,
	displayLabel: true,
	displayTimestamp: true,
	underlineLabel: true,
	underlineMessage: false,
	underlinePrefix: false,
	underlineSuffix: false,
	uppercaseLabel: false,
})

export * from '@/utils/Errors'
export * from '@/utils/testing/supertest'

export {
	Errors,
	RD,
	signale,
	log,
	bb,
	flattenGQLResponse,
	warn,
	sleep,
	consoleWrite,
	gql
}
