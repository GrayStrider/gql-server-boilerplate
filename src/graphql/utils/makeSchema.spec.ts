import {makeSchema} from '@/graphql/utils/index'
import {promises as fs} from 'fs'
import {join} from 'path'


describe('make schema', () => {
  it('should create schema', async () => {
    expect.assertions(1)
    const schema = `
    type Query {
      foo: String
    }
    `
    const path = join(__dirname, 'schema.graphql')
    await fs.writeFile(path, schema)
    
    const res = await makeSchema(path, {})
    expect(res).toMatchSnapshot()
    await fs.unlink(path)
  })
  
})
