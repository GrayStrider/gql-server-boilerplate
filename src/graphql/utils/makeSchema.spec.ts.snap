// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`make schema should create schema 1`] = `
GraphQLSchema {
  "__allowedLegacyNames": Array [],
  "__validationErrors": undefined,
  "_directives": Array [
    "@skip",
    "@include",
    "@deprecated",
  ],
  "_implementations": Object {},
  "_mutationType": undefined,
  "_possibleTypeMap": Object {},
  "_queryType": "Query",
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Query": "Query",
    "String": "String",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": undefined,
  "extensionASTNodes": undefined,
  "extensions": undefined,
}
`;
